lib/player.js

```diff
@@ -1,19 +1,99 @@
+/* eslint-env node */
+
+const uuidV4 = require('uuid/v4');
+
+const Game = require('./game');
+
+
 class Player {
-  constructor(socket) {
+  constructor(server, socket) {
+    this.id = uuidV4();
+    this.server = server;
     this.socket = socket;
     this.game = null;
-    socket.on('game event', (data) => {
-      if (this.game) {
-        this.game.addEvent(data, this);
-      }
+    this.installEventListeners();
+  }
+
+  installEventListeners() {
+    this.socket.on('disconnect', () => {
+      this.server.removePlayer(this.id);
+    });
+
+    this.socket.on('game list', () => {
+      this.socket.emit('game list', {
+        games: Object.keys(this.server.games).map((id) => {
+          const game = this.server.games[id];
+
+          return {
+            id,
+            mode: game.mode,
+            playerCount: game.playerCount,
+            maximumPlayerCount: game.maximumPlayerCount,
+          };
+        }),
+      });
+    });
+
+    this.socket.on('game create', (data) => {
+      this.createGame(data.mode);
     });
-    socket.on('disconnect', () => {
-      process.stdout.write(`${this.socket.conn.remoteAddress} disconnected\n`);
+
+    this.socket.on('game join', (data) => {
+      this.joinGame(data.id);
+    });
+
+    this.socket.on('game event', (data) => {
       if (this.game) {
-        this.game.removePlayer(this);
+        this.game.addEvent(data, this);
       }
     });
   }
+
+  createGame(gameModeName) {
+    // Currently we support only game mode called 'vs', even though that is not
+    // very descriptive name.
+    if (gameModeName !== 'vs') {
+      this.socket.emit('client error', { message: `Unrecognized game mode: ${gameModeName}` });
+      return;
+    }
+
+    // Players shouldn't be able to create more games if they have already been
+    // joined to another one.
+    if (this.game) {
+      this.socket.emit('client error', { message: 'Already joined in a game' });
+      return;
+    }
+
+    this.game = new Game(this);
```

Comment:
This also causes a stuttering effect because `server.frameRate` isn't passed correctly on connection.